### PR TITLE
Reuse byte slices from relay handler

### PR DESCRIPTION
**Pull request by ryanslade**
When using the relay handler the previous code would:

Create a empty byte buffer
Use it to build up the response (as raw json)
Return the raw bytes, converted to a json.RawMessage
Marshal the RawMessage which simply returns the bytes unhanged
Write the bytes to the a ResponseWriter
Throw away the bytes

The new code instead passes the final bytes back into a buffer pool so
that they can be reused when handling subsequent requests.

### CODE REVIEW

**Comment by ryanslade** [comment]
This feels like a hack because I needed to expose the byte pool, but this accounts for a huge amount of allocations in frontend so I think it might be worth a try.

File: internal/exec/exec.go [hunk 1]
```diff
@@ -72,21 +88,29 @@ func (r *Request) execSelections(ctx context.Context, sels []selected.Selection,
...
+				BytePool.Put(f.out.Bytes())
+			}
+		}
+	}()
+
 	if async {
 		var wg sync.WaitGroup
 		wg.Add(len(fields))
 		for _, f := range fields {
 			go func(f *fieldToExec) {
```
Comment (keegancsmith) [inline internal/exec/exec.go:103]: I remember us adding some code here for bounding the number of goroutines. What happened to it?

**Review by keegancsmith — APPROVED** [review]
This does indeed feel very hacky. I'm not sure how much benefit one gets from bytepools these days. It would be interesting to see how much of a difference this makes.

One observation I have reading this code is we don't need to generate the full list of `.out` since we just loop over it serialling once they are all generated. Maybe if we had a nice way to concurrently run over the list but return results in the same order we could do something like that.

File: internal/exec/exec.go (hunk 1 above)
Comment (ryanslade) [reply internal/exec/exec.go:103]: Our change only limited the number of list items we handle concurrently, not fields:
https://github.com/sourcegraph/graphql-go/blob/e67ac72893d245ce10092afba20bd2289a2c31c1/internal/exec/exec.go#L341

**Comment by ryanslade** [comment]
I did some testing locally and although this did reduce the number of allocations I don't think it's worth the hack. We still end up with a lot of byte allocations due to json encoding in many places and I think a better long term option is actually going to be to investigate some other graphql libraries.

**Pull request closed without merging**